definitions/npm/styled-components_v4.x.x/flow_v0.92.0-/styled-components_v4.x.x.js

```diff
@@ -0,0 +1,431 @@
+    css(): string;
+    toHTML(additionalAttrs: ?string): string;
+    toElement(): React$Element<*>;
+    clone(): Tag<T>;
+    sealed: boolean;
+  }
+
+  // The `any`/weak types in here all come from `styled-components` directly, since those definitions were just copied over
+  declare export class StyleSheet {
+    static get master() : StyleSheet;
+    static get instance() : StyleSheet;
+    static reset(forceServer? : boolean) : void;
+
+    id : number;
+    forceServer : boolean;
+    target : ?HTMLElement;
+    tagMap : {[string]: Tag<any>}; // eslint-disable-line flowtype/no-weak-types
+    deferred: { [string]: string[] | void };
+    rehydratedNames: { [string]: boolean };
+    ignoreRehydratedNames: { [string]: boolean };
+    tags: Tag<any>[]; // eslint-disable-line flowtype/no-weak-types
+    importRuleTag: Tag<any>; // eslint-disable-line flowtype/no-weak-types
+    capacity: number;
+    clones: StyleSheet[];
+
+    constructor(?HTMLElement) : this;
+    rehydrate() : this;
+    clone() : StyleSheet;
+    sealAllTags() : void;
+    makeTag(tag : ?Tag<any>) : Tag<any>; // eslint-disable-line flowtype/no-weak-types
+    getImportRuleTag() : Tag<any>; // eslint-disable-line flowtype/no-weak-types
+    getTagForId(id : string): Tag<any>; // eslint-disable-line flowtype/no-weak-types
+    hasId(id: string) : boolean;
+    hasNameForId(id: string, name: string) : boolean;
+    deferredInject(id : string, cssRules : string[]) : void;
+    inject(id: string, cssRules : string[], name? : string) : void;
+    remove(id : string) : void;
+    toHtml() : string;
+    toReactElements() : React$ElementType[];
+  }
+
+  declare export class KeyFrames {
+    id : string;
+    name : string;
+    rules : string[];
+
+    constructor(name : string, rules : string[]) : this;
+    inject(StyleSheet) : void;
+    toString() : string;
+    getName() : string;
+  }
+
+  // I think any is appropriate here?
+  // eslint-disable-next-line flowtype/no-weak-types
+  declare export var css : CSSConstructor;
+  declare export var keyframes : KeyFramesConstructor;
+  declare export var createGlobalStyle : CreateGlobalStyleConstructor
+  declare export var ThemeProvider: React$ComponentType<{children?: ?React$Node, theme: mixed | (mixed) => mixed}>
+
+  // This is a bit hard to read. Not sure how to make it more readable. I think adding line-breaks makes it worse.
+  declare type ThemeProps<T> = {
+    theme: T
+  }
+
+  declare export function withTheme<Theme, Config: {}, Instance>(Component: React$AbstractComponent<Config, Instance>): React$AbstractComponent<$Diff<Config, ThemeProps<Theme | void>>, Instance>
+
+  declare export type StyledComponent<Props, Theme, Instance> = React$AbstractComponent<Props, Instance> & InterpolatableComponent
+  declare interface StyledFactory<StyleProps, OwnProps, Theme, Instance> {
+    [[call]]: TaggedTemplateLiteral<StyleProps & OwnProps & ThemeProps<Theme>, StyledComponent<StyleProps & OwnProps, Theme, Instance>>;
```

Comment:
@omninonsense, I'm looking into options for replacing the intersections here with object spread instead. As I do that, I have a question about the design intent here:

It looks like `StyleProps` here is intended to only be the props used within the styled-component interpolations, is that correct? Would it make sense for us to pass in `StyleProps` as _all_ of the props that the styled component exports? As an example of how this could be used:

```js
type Props = {name : string}
  type StyleProps = { ...Props, color?: string}
  type Theme = {
    accent: string
  }

  const Hello = (p: Props) =>
    <div>Hello {p.name}</div>

  const StyledHello = styled<StyleProps, Theme, Props, *>(Hello)`
    color: ${props => props.color || props.theme.accent};
  `
```

At first glance it looks like this might simplify implementation a bit, as well as be simpler for consumers to maintain (they wouldn't have to update the `StyleProps` if they wanted to use an existing `Prop` in an interpolation). Thoughts?

Reply:
I like your idea better. I can't remember where I was going with the original idea!

Reply:
> When testing this out on an existing codebase I'm getting a number of the following errors:
> 
> ```
> Missing type annotation for SP. SP is a type parameter declared in function type [1] and was implicitly instantiated
> at call of styled [2].
> 
> Missing type annotation for Th. Th is a type parameter declared in function type [1] and was implicitly instantiated
> at call of styled [2].
> ```
> 
> These errors make sense with the new changes, but I wonder if there are opportunities for us to improve the developer experience for people who are suddenly going to see a ton of these errors. A first order change may be to update the names (e.g. `SP` to `StyledProps`, and `Th` to `Theme`), so that people have a better understanding of what the error is complaining about. Ideally, though, we would also be able to set defaults for these arguments, so that Flow complains about the errors at the site of the interpolation instead. As far as I'm aware, though, there's no way to add default arguments in lib defs. @omninonsense, do you know of any workarounds?

I don't think there are any workarounds. I guess passing in an empty object might work (haven't tried)… Although, I am a bit reluctant on adding that in, since I am not sure if it will contribute much. IMO, people can always make their own utility types to help with these. Something along these lines:

~~~
type Theme = {| … |}
type MySC<Props> = StyledComponent<Props, Theme, *>
~~~

and they should be able to use it as `const msc: MySC<SomeProps> = styled(…)`

Reply:
@edahlseng Implemented this in 45f8151. Thoughts?

Reply:
**Note**: This also switched to using spread from using intersections, so we should have less "surprises"